src/renderer/qss/js/qssFocusManager.js

```diff
@@ -0,0 +1,331 @@
+            nextElementIndex = focusGroupInfo.focusIndex + delta;
+
+        if (focusGroupIndex > -1) {
+            while (0 <= nextElementIndex && nextElementIndex < focusGroup.length) {
+                var elementToFocus = focusGroup[nextElementIndex];
+                if (that.isFocusable(elementToFocus)) {
+                    that.focusElement(elementToFocus, true);
+                    break;
+                } else {
+                    nextElementIndex += delta;
+                }
+            }
+        }
+    };
+
+    /**
+     * Focuses the first focusable button which comes before the currently focused button (if any)
+     * in the same focus group. If there is no focused group initially, this function does nothing.
+     * @param {Component} that - The `gpii.qss.qssFocusManager` instance.
+     */
+    gpii.qss.qssFocusManager.onArrowUpPressed = function (that) {
+        gpii.qss.qssFocusManager.focusNearestVertically(that, false);
+    };
+
+    /**
+     * Focuses the first focusable button which comes after the currently focused button (if any)
+     * in the same focus group. If there is no focused group initially, this function does nothing.
+     * @param {Component} that - The `gpii.qss.qssFocusManager` instance.
+     */
+    gpii.qss.qssFocusManager.onArrowDownPressed = function (that) {
+        gpii.qss.qssFocusManager.focusNearestVertically(that, true);
+    };
+
+    /**
+     * Focuses the first available button which conforms to all of the following conditions:
+     * 1. The button is focusable.
+     * 2. The button is in the first focusable group which comes to the left or to the right
+     * (depending on the value of the `direction` flag) of the current focus group. If there
+     * is no such button in any of the groups to the left or to the right, the groups to the right
+     * or to the left of the current focus group respectively are also examined starting from the
+     * farthest.
+     * 3. The button has the same index in its focus group as the index of the currently focused
+     * button in its group. If the new group has fewer buttons, its last button will be focused.
+     * @param {Component} that - The `gpii.qss.qssFocusManager` instance.
+     * @param {FocusGroupsInfo} focusGroupInfo - An object holding information about the focusable
+     * elements in the QSS.
+     * @param {Number} initialGroupIndex - the index of the focus group from which the examination
+     * should commence
+     * @param {Boolean} direction - If `true` the scanning direction will be from left to right.
+     * Otherwise, it will be from right to left.
+     */
+    gpii.qss.qssFocusManager.focusNearestHorizontally = function (that, focusGroupInfo, initialGroupIndex, direction) {
+        var focusGroups = focusGroupInfo.focusGroups,
+            focusIndex = focusGroupInfo.focusIndex,
+            delta = direction ? 1 : -1,
+            nextGroupIndex = initialGroupIndex;
+
+        do {
```

Comment:
could we use only a `while` here?

Reply:
Basically yes but this works well for handling edge cases (e.g. when there is no focus group initially). That is why I chose this approach.